plat/common/arm/lcpu.c

```diff
@@ -0,0 +1,250 @@
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the copyright holder nor the names of its
+ *    contributors may be used to endorse or promote products derived from
+ *    this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <stdint.h>
+#include <uk/plat/lcpu.h>
+#include <uk/plat/io.h>
+#include <arm/smccc.h>
+#include <arm/arm64/cpu.h>
+#include <gic/gic.h>
+#include <libfdt.h>
+#include <ofw/fdt.h>
+#include <stdint.h>
+
+/**
+ * The number of cells for the size field should be 0 in cpu nodes.
+ * The number of cells in the address field is set by default to 2 in cpu
+ * nodes. This is also the maximum number of cells for the address field that
+ * we can currently support.
+ **/
+#define FDT_SIZE_CELLS_DEFAULT 0
+#define FDT_ADDR_CELLS_DEFAULT 2
+#define CPU_ID_MASK 0xff00ffffffUL
+
+static void *dtb;
+void lcpu_start(struct lcpu *cpu);
+static __paddr_t lcpu_start_paddr;
+extern struct _gic_dev *gic;
+
+__lcpuid lcpu_arch_id(void)
+{
+	__u64 mpidr_reg;
+
+	mpidr_reg = SYSREG_READ64(mpidr_el1);
+
+	/* return the affinity bits for the current core */
+	return mpidr_reg & CPU_ID_MASK;
+}
+
+#ifdef CONFIG_HAVE_SMP
+int lcpu_arch_init(struct lcpu *this_lcpu)
+{
+	int ret = 0;
+
+	/* Initialize the interrupt controller for non-bsp cores */
+	if (!lcpu_is_bsp(this_lcpu)) {
+		ret = gic->ops.initialize();
+		if (unlikely(ret))
+			return ret;
+	}
+
+	return ret;
+}
+
+void __noreturn lcpu_arch_jump_to(void *sp, ukplat_lcpu_entry_t entry)
+{
+	__asm__ __volatile__ (
+		"mov	sp, %0\n" /* set the sp  */
+		"br	%1\n"     /* branch to the entry function */
+		:
+		: "r"(sp), "r"(entry)
+		: "sp");
+
+	/* just make the compiler happy about returning function */
+	ukplat_lcpu_halt();
```

Comment:
Just as an explanation for why I recommended a `__builtin_unreachable()`. The latter will not emit any code but just make the compiler happy. While adding the call to halt will add code that is never executed.
```ASM
void __noreturn lcpu_arch_jump_to(void *sp, ukplat_lcpu_entry_t entry)
{
        __asm__ __volatile__ (
    40100f50:   9100001f        mov     sp, x0
    40100f54:   d61f0020        br      x1

0000000040100f58 <lcpu_arch_mp_init>:
        __builtin_unreachable();
}
```
vs.
```ASM
void __noreturn lcpu_arch_jump_to(void *sp, ukplat_lcpu_entry_t entry)
{
    40100f50:   a9bf7bfd        stp     x29, x30, [sp, #-16]!
    40100f54:   910003fd        mov     x29, sp
        __asm__ __volatile__ (
    40100f58:   9100001f        mov     sp, x0
    40100f5c:   d61f0020        br      x1
                :
                : "r"(sp), "r"(entry)
                : /* sp not needed */);

        /* just make the compiler happy about returning function */
        ukplat_lcpu_halt();
    40100f60:   94000c96        bl      401041b8 <ukplat_lcpu_halt>
    40100f64:   d503201f        nop
0000000040100f68 <lcpu_arch_mp_init>:
}
```